hub/hub.go

```diff
@@ -4,80 +4,186 @@ import (
-func (h *Hub) Stop() error {
-	return h.transport.Close()
+// WithDebug enables the debug mode.
+func WithDebug() Option {
+	return func(o *Hub) {
+		o.debug = true
+	}
 }
 
-// NewHub creates a hub using the Viper configuration.
-func NewHub(v *viper.Viper) (*Hub, error) {
-	if err := ValidateConfig(v); err != nil {
-		return nil, err
+// WithDemo enables the demo mode.
+func WithDemo() Option {
+	return func(h *Hub) {
+		h.demo = true
 	}
+}
 
-	var (
-		logger Logger
-		err    error
-	)
+// WithMetrics enables collection of Prometheus metrics.
+func WithMetrics() Option {
+	return func(h *Hub) {
+		h.metrics = NewMetrics()
+	}
+}
 
-	if v.GetBool("debug") {
-		logger, err = zap.NewDevelopment()
-	} else {
-		logger, err = zap.NewProduction()
+// WithSubscriptions allows to dispatch updates when subscriptions are created or terminated.
+func WithSubscriptions() Option {
+	return func(h *Hub) {
+		h.subscriptions = true
 	}
-	if err != nil {
-		return nil, fmt.Errorf("can't initialize zap logger: %w", err)
+}
+
+// WithLogger sets the logger to use.
+func WithLogger(logger Logger) Option {
+	return func(h *Hub) {
+		h.logger = logger
 	}
+}
 
-	t, err := NewTransport(v, logger)
-	if err != nil {
-		return nil, err
+// WithWriteTimeout sets maximum duration before closing the connection, defaults to 600s, set to 0 to disable.
+func WithWriteTimeout(timeout time.Duration) Option {
+	return func(h *Hub) {
+		h.writeTimeout = timeout
+	}
+}
+
+// WithDispatchTimeout sets the maximum dispatch duration of an update.
+func WithDispatchTimeout(timeout time.Duration) Option {
+	return func(h *Hub) {
+		h.dispatchTimeout = timeout
+	}
+}
+
+// WithHeartbeat enables heartbeat.
+func WithHeartbeat(interval time.Duration) Option {
+	return func(h *Hub) {
+		h.heartbeat = interval
 	}
+}
 
-	return NewHubWithTransport(v, t, logger, NewTopicSelectorStore()), nil
+// WithPublisherJWTConfig sets the JWT key and algorithm to use.
+func WithPublisherJWTConfig(key []byte, signingMethod jwt.SigningMethod) Option {
+	return func(h *Hub) {
+		h.publisherJWTConfig = &jwtConfig{key, signingMethod}
+	}
+}
+
+// WithSubscriberJWTConfig sets the JWT key and algorithm to use.
+func WithSubscriberJWTConfig(key []byte, signingMethod jwt.SigningMethod) Option {
+	return func(h *Hub) {
+		h.subscriberJWTConfig = &jwtConfig{key, signingMethod}
+	}
 }
 
-// NewHubWithTransport creates a hub.
-func NewHubWithTransport(v *viper.Viper, t Transport, logger Logger, tss *TopicSelectorStore) *Hub {
-	return &Hub{
-		v,
-		logger,
-		t,
-		nil,
-		nil,
-		tss,
-		NewMetrics(),
+// WithPublishOrigins sets the origins allowed to publish updates.
+func WithPublishOrigins(origins []string) Option {
+	return func(h *Hub) {
+		h.publishOrigins = origins
 	}
 }
 
-// Start is an helper method to start the Mercure Hub.
-func Start() {
-	h, err := NewHub(viper.GetViper())
+// WithTransportURL sets the transport to use by parsing the provided URL.
+func WithTransportURL(tu string) Option {
```

Comment:
TBH, I like this approach even less. IMHO the Hub itself should not be concerned about how a transport is constructed. It should rely solely on the `Transport` interface. Forcing to use a URL makes the library less reusable in my opinion.

Reply:
I'll write the Caddy module and it will help to see the use case and if a refactoring is necessary.

Reply:
Consider the following use case: I want to integrate mercure into an existing server process (because I don't need microservices, I'm perfectly fine with a single process). I already have a postgres connection open for my application, I want to reuse it for my mercure transport as well.

IMHO that's a better use case showcasing the reusability of mercure as a library than the Caddy module.

Reply:
@sagikazarmark this use case is already supported. Just register your own factory:

```go
func main() {
        pgCon := ...
	mercure.RegisterTransportFactory("pg", func newBoltTransport(u *url.URL, l Logger) (Transport, error) {
            // use pgCon here
        })
       h := mercure.NewHub(mercure.WithTransport(&url.URL{Scheme: "pg"}))
       defer h.Stop();
       // register the Mercure handler in your server
}
```

The current approach as another benefit. If you create a package calling `RegisterTransportFactory` in an `init()` function, the transport is automatically available in the Caddy module (by passing the appropriate `transport_url`in the `Caddyfile`): https://github.com/dunglas/mercure/pull/393/files#diff-be49a2dcca4004e34971b1e8dedc623412b90f41fbc7c18e1488cb73c73ec90fR96

As you probably know, I sell extra transports as part of the "on promise" version, so this feature is very important for me, but actually this will be helpful to everybody: create your transport, run:

```
xcaddy build \
      --with github.com/user/your-transport \             
      --with github.com/dunglas/mercure
```

And you have a Caddy build supporting Mercure and using your own transport!

This can definitely be improved (for instance the transport registry could be in another module, and/or we can expose a `WithTransport` function to bypass the factory if necessary), but IMHO it can be done in a next step.

WDYT?

Reply:
> this use case is already supported. Just register your own factory:

I think registering a global transport factory with a local DB connection is a bad idea and is unnecessarily complicated.

> the transport is automatically available in the Caddy module

I don't see why the two can't co-exist though. Having a global factory registry doesn't mean that you have to call the factory inside `Hub`. IMHO the caddy module should call it and register the `Close` method in a shutdown hook instead of calling it in `Hub.Stop`.

> This can definitely be improved (for instance the transport registry could be in another module, and/or we can expose a WithTransport function to bypass the factory if necessary), but IMHO it can be done in a next step.

Yeah, it can definitely be part of a separate PR, this is quite huge already. 🙂

Reply:
Refactored! Thanks for the feedback.

Reply:
Looks great, thanks!